csi/moac/volume.ts

```diff
@@ -0,0 +1,877 @@
+  //
+  constructor(
+    uuid: string,
+    registry: any,
+    emitEvent: (type: string) => void,
+    spec: any,
+    state?: VolumeState,
+    size?: number,
+    publishedOn?: string,
+  ) {
+    assert(spec);
+    // specification of the volume
+    this.uuid = uuid;
+    this.registry = registry;
+    this.replicaCount = spec.replicaCount || 1;
+    this.preferredNodes = _.clone(spec.preferredNodes || []).sort();
+    this.requiredNodes = _.clone(spec.requiredNodes || []).sort();
+    this.requiredBytes = spec.requiredBytes;
+    this.limitBytes = spec.limitBytes;
+    this.protocol = spec.protocol;
+    // state variables of the volume
+    this.size = size || 0;
+    this.publishedOn = publishedOn;
+    this.nexus = null;
+    this.replicas = {};
+    this.state = state || VolumeState.Pending;
+    // other properties
+    this.runFsa = 0;
+    this.nodeBlackList = {};
+    this.emitEvent = emitEvent;
+  }
+
+  // Stringify volume
+  toString(): string {
+    return this.uuid;
+  }
+
+  // Get the size of the volume.
+  getSize(): number {
+    return this.size;
+  }
+
+  // Get the node where the volume is accessible from (that is the node with
+  // the nexus) or undefined when nexus does not exist (unpublished/published).
+  getNodeName(): string | undefined {
+    return this.publishedOn;
+  }
+
+  // Publish the volume. That means, make it accessible through a block device.
+  //
+  // @params protocol      The nexus share protocol.
+  // @return uri           The URI to access the nexus.
+  async publish(protocol: Protocol): Promise<string> {
+    if (this.state !== VolumeState.Degraded && this.state !== VolumeState.Healthy) {
+      throw new GrpcError(
+        GrpcCode.INTERNAL,
+        'Cannot publish a volume that is neither healthy nor degraded'
+      );
+    }
+    let nexus = this.nexus;
+    if (!nexus) {
+      // Ensure replicas can be accessed from nexus. Set share protocols.
+      const [nexusNode, replicaSet] = await this._ensureReplicaShareProtocols();
+
+      if (!this.size) {
+        // the size will be the smallest replica
+        this.size = Object.values(this.replicas)
+          .map((r) => r.size)
+          .reduce((acc, cur) => (cur < acc ? cur : acc), Number.MAX_SAFE_INTEGER);
+      }
+      // create a new nexus with children (replicas) created in previous steps
+      nexus = await this._createNexus(nexusNode, replicaSet);
+    } else {
+      log.debug(`Publishing volume ${this} that already has a nexus`)
+    }
+    let uri = nexus.getUri();
+    if (!uri) {
+      uri = await nexus.publish(protocol);
+    } else {
+      log.debug(`Publishing volume ${this} that has been already published`)
+    }
+    this.publishedOn = nexus.node.name;
+    log.info(`Published "${this}" at ${uri}`);
+    this.emitEvent('mod');
+    return uri;
+  }
+
+  // Undo publish operation on the volume.
+  async unpublish() {
+    if (this.publishedOn) {
+      this.publishedOn = undefined;
```

Comment:
some of the code here will need to change for sure. Currently we destroy nexus no matter what upon unpublish. The problem with unsetting publishedOn only when the nexus is destroyed is that then moac does not know that it should destroy the nexus in FSA. So we would have to introduce another field to distinguish published state from still-published-but-should-not-be and that to me sounds like an overkill. The nexus will be removed at first opportunity when it's possible so unsetting publishedOn and returning OK for the operation in spite of the fact that the nexus might still be there seems ok to me. Also in respect to what might happen next - that the volume may be published again - we just use the nexus that is there, so should not cause any harm.

Reply:
hmm, what if the destroy fails for some other reason (bug on mayastor or something). Once the user pod runs again is there a possibility that we would try to create the nexus on a different node? (at that point we'd have 2 of the same nexus on different nodes)

Reply:
this goes a bit into the deferred nexus destruction feature that I'm going to implement in the future, but to me it seems it is not possible. If there is a next request to publish a volume then the nexus we failed to destroy is still there, so we will use it. If the nexus is unusable for any reason then application will probably notice that as it will receive IO errors.